src/actions/mintArveaweNFT.ts

```diff
@@ -0,0 +1,266 @@
+  ]);
+  const files = new Map<string, Buffer>();
+  files.set(file.name, dataFile);
+  files.set(fileMetadata.name, dataMetadata);
+  return getAssetCostInfo(files);
+}
+export async function getCost(
+  file: File,
+  metadata: MetadataJson,
+  fileMetadata = createFilePack(metadata),
+) {
+  const calcInfo = await getCostInfo(file, metadata, fileMetadata);
+  return calcInfo.solana * LAMPORTS_PER_SOL;
+}
+
+const ARWEAVE_WALLET = new PublicKey('6FKvsq4ydWFci6nGq9ckbjYMtnmaqAoatz5c9XWjiDuS');
+
+export async function payForFiles(
+  {
+    file,
+    metadata,
+    wallet: w,
+  }: {
+    wallet: Wallet;
+    file: File;
+    metadata: MetadataJson;
+  },
+  WebFile = File,
+) {
+  const fileMetadata = createFilePack(metadata, WebFile);
+  const [hashFile, hashMetadata] = await Promise.all([
+    convertFile2Hash(file),
+    convertFile2Hash(fileMetadata),
+  ]);
+  const files = new Map<string, File>();
+  files.set(file.name, file);
+  files.set(fileMetadata.name, fileMetadata);
+
+  const lamports = await getCost(file, metadata, fileMetadata);
+
+  const payForFilesTx = new PayForFiles(
+    {
+      feePayer: w.publicKey,
+    },
+    {
+      arweaveWallet: ARWEAVE_WALLET,
+      lamports,
+      fileHashes: [hashFile, hashMetadata],
+    },
+  );
+
+  return { payForFilesTx, files };
+}
+
+export interface MintArveaweNFTResponse {
+  txId: string;
+  mint: PublicKey;
+  metadata: PublicKey;
+  arweaveResult: ArweaveUploadResult;
+}
+
+export async function mintArveaweNFT(
+  {
+    connection,
+    wallet,
+    file,
+    metadata,
+    maxSupply,
+    storage,
+    progress = () => {},
+  }: {
+    connection: Connection;
+    wallet: Wallet;
+    storage: ArweaveStorage;
+    file: File;
+    metadata: MetadataJson;
+    maxSupply: number;
+    progress?: (status: ENftProgress) => void;
+  },
+  WebFile = File,
+): Promise<MintArveaweNFTResponse> {
+  try {
+    progress(ENftProgress.minting);
+
+    const { payForFilesTx, files } = await payForFiles(
+      {
+        wallet,
+        file,
+        metadata,
+      },
+      WebFile,
+    );
+
+    const { mint, createMintTx, createAssociatedTokenAccountTx, mintToTx } =
+      await prepareTokenAccountAndMintTx(connection, wallet.publicKey);
+
+    const metadataPDA = await Metadata.getPDA(mint.publicKey);
+
+    const creators = metadata.properties.creators.map((c) => new Creator(c));
+    const metadataData = new MetadataDataData({
+      name: metadata.name,
+      symbol: metadata.symbol,
+      uri: ' '.repeat(64),
+      sellerFeeBasisPoints: metadata.seller_fee_basis_points,
+      creators,
+    });
+    const createMetadataTx = new CreateMetadata(
+      {
+        feePayer: wallet.publicKey,
+      },
+      {
+        metadata: metadataPDA,
+        mint: mint.publicKey,
+        mintAuthority: wallet.publicKey,
+        updateAuthority: wallet.publicKey,
+        metadataData,
+      },
+    );
+    progress(ENftProgress.preparing_assets);
```

Comment:
Is there really a point to update progress on non async executions?

Reply:
Yes. This operation keeps a lot of time and we show this progress in UI